### PR TITLE
Feature/extract/policy fragments

### LINKED ISSUE #736: supporting the new "Policy fragments" feature in the extractor
Do you have roadmap for adding support for the new feature of apim  "Policy fragments" to the extractor

### PR DESCRIPTION
**Pull request by f-alizada**
Add policy fragment extraction. 
Add generated template to main linked template
Closes #736

### CODE REVIEW

File: tests/ArmTemplates.Tests/Moqs/ApiClients/MockPolicyFragmentClient.cs [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+using Microsoft.Azure.Management.ApiManagement.ArmTemplates.Extractor.Utilities.DataProcessors;
+using Moq;
+
+namespace Microsoft.Azure.Management.ApiManagement.ArmTemplates.Tests.Moqs.ApiClients
+{
+    class MockPolicyFragmentClient
+    {
+        public static async Task<IPolicyFragmentsClient> GetMockedHttpPolicyFragmentClient(string responseFileLocation)
+        {
+            var mockedProcessor = new PolicyFragmentDataProcessor();
```
[the remark this answers is not in the document]
Comment (f-alizada) [reply tests/ArmTemplates.Tests/Moqs/ApiClients/MockPolicyFragmentClient.cs:18]: I was thinking to remove mocks for dataProcessors where it makes sense, since dataprocessor is an additional step in processing data, it allows us to catch the error on development stage if something would go wrong. 
Apart from that the tests that uses this mockedHttpClient should go from fetching the results till the generation of the template. 
What do you think?

File: src/ArmTemplates/Extractor/EntityExtractors/PolicyFragmentsExtractor.cs [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+                policyFragment.ApiVersion = GlobalConstants.ApiVersionPreview;
+
+                if (string.IsNullOrEmpty(extractorParameters.SingleApiName))
+                {
+                    policyFragmentsTemplate.TypedResources.PolicyFragments.Add(policyFragment);
+                }
+                else
+                {
+                    var policyFragmentReferenceString = $"fragment-id=\"{policyFragment.OriginalName}\"";
+                    var isReferencedInPolicy = apiTemplatePolicies?.Any(x => x.Properties.PolicyContent.Contains(policyFragmentReferenceString));
```
[the remark this answers is not in the document]
Comment (f-alizada) [reply src/ArmTemplates/Extractor/EntityExtractors/PolicyFragmentsExtractor.cs:64]: There is existing api endpoint: https://docs.microsoft.com/en-us/rest/api/apimanagement/current-preview/policy-fragment/list-references?tabs=HTTP which allows us to fetch all resources where policyFragment is connected. 
However, in case when there will be many policy fragments, we should make additional API call for each of them to check if it is referenced or not. This is good from one source-of-truth point but adds additional extraction time (depends of course on the count of policy fragments) 
Having that the policy fragment when referenced should have $"fragment-id=\"{policyFragment}\ this exact string in the policy, apart from that to have more accurate results there is a possibility to parse the xml and verify that it is referenced using include-fragment xml element. 
Basically, there are two options: 
1 - add additional api call for each policy fragment -> which increases time of exporting the templates. 
2 - add additional xml element check 
Open to discussion

[the remark this answers is not in the document]
Comment (f-alizada) [reply src/ArmTemplates/Extractor/EntityExtractors/PolicyFragmentsExtractor.cs:64]: I do agree with the source of truth to be one api call, but image scenario when one api is being extracted.
There 1000 policy fragments -> 1000 additional api calls having Api policy content is extracted and available for comparison without additional call. 
Regarding error prone: the search pattern was built using the documentation of policy-fragment resource, worst case scenario is when policy fragment will be extracted that is not used by API, in case there will be some custom xml which contains exact same value.

**Review by DeagleGross — APPROVED** [review]
